File: src/main/java/com/smarthome/backend/server/api/modules/hue/HueModuleManager.java
package com.smarthome.backend.server.api.modules.hue;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.Gson;
import com.smarthome.backend.model.devices.Device;
import com.smarthome.backend.model.devices.DeviceLight;
import com.smarthome.backend.model.devices.DeviceLightDimmer;
import com.smarthome.backend.model.devices.DeviceLightDimmerTemperature;
import com.smarthome.backend.model.devices.DeviceLightDimmerTemperatureColor;
import com.smarthome.backend.model.devices.DeviceMotion;
import com.smarthome.backend.server.actions.ActionManager;
import com.smarthome.backend.server.api.ApiRouter;
import com.smarthome.backend.server.api.modules.ModuleManager;
import com.smarthome.backend.server.db.DatabaseManager;
import com.smarthome.backend.server.events.EventStreamManager;

/**
 * Modul-Manager für Hue Bridge-Verwaltung.
 * Bietet Funktionen zur Bridge-Entdeckung und -Verwaltung.
 */
public class HueModuleManager extends ModuleManager {
    private static final Logger logger = LoggerFactory.getLogger(HueModuleManager.class);
    private static final Gson gson = new Gson();
    
    private final HueDiscover hueDiscover;
    private final HueBridgeController hueBridgeController;
    private final HueDeviceController hueDeviceController;
    
    public HueModuleManager(DatabaseManager databaseManager, EventStreamManager eventStreamManager, ActionManager actionManager) {
        super(databaseManager, eventStreamManager, actionManager);
        this.hueDiscover = new HueDiscover(databaseManager);
        this.hueBridgeController = new HueBridgeController(databaseManager);
        this.hueDeviceController = new HueDeviceController(databaseManager);
        this.setHueDeviceController();
        eventStreamManager.registerModuleEventStreamManager(hueBridgeController.createEventStreamManager(actionManager));
    }

    public void setHueDeviceController() {
        List<Device> devices = actionManager.getDevices();
        for(Device device : devices) {
            this.setHueDeviceControllerForDevice(device);
        }
    }

    public void setHueDeviceControllerForDevice(Device device) {
        if (device instanceof HueLight) {
            ((HueLight) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueLightDimmer) {
            ((HueLightDimmer) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueLightDimmerTemperatureColor) {
            ((HueLightDimmerTemperatureColor) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueLightDimmerTemperature) {
            ((HueLightDimmerTemperature) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueLightLevelSensor) {
            ((HueLightLevelSensor) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueTemperatureSensor) {
            ((HueTemperatureSensor) device).setHueDeviceController(hueDeviceController);
        } else if (device instanceof HueMotionSensor) {
            ((HueMotionSensor) device).setHueDeviceController(hueDeviceController);
        }
    }
    
    /**
     * Sucht nach verfügbaren Hue Bridges im Netzwerk.
     * Gefundene Bridges werden zurückgegeben.
     * 
     * @param exchange HTTP-Exchange für die Antwort
     * @throws IOException bei Fehlern beim Senden der Antwort
     */
    public void discoverBridges(com.sun.net.httpserver.HttpExchange exchange) throws IOException {
        logger.info("Suche nach Hue Bridges");
        
        try {
            // Führe Discovery durch (5 Sekunden)
            long searchDurationMs = 5000;
            List<HueDiscoveredBridge> bridges = hueDiscover.discoverBridges(searchDurationMs);
            
            logger.info("{} Bridges gefunden", bridges.size());
            
            // Entferne sensible Daten (username, clientKey) vor der Serialisierung
            List<HueDiscoveredBridge> bridgesForApi = new java.util.ArrayList<>();
            for (HueDiscoveredBridge bridge : bridges) {
                bridgesForApi.add(bridge.withoutSensitiveData());
            }
            
            // Sende Antwort
            String response = gson.toJson(bridgesForApi);
            ApiRouter.sendResponse(exchange, 200, response);
            
        } catch (Exception e) {
            logger.error("Fehler bei der Bridge-Erkennung", e);
            ApiRouter.sendResponse(exchange, 500, 
                gson.toJson(Map.of("error", "Fehler bei der Bridge-Erkennung: " + e.getMessage())));
        }
    }
    
    /**
     * Führt das Pairing mit einer Hue Bridge durch.
     * Delegiert den Aufruf an HueBridgeController und registriert den Event-Stream-Manager nach erfolgreichem Pairing.
     * 
     * @param exchange HTTP-Exchange für die Antwort
     * @param bridgeId Die deviceId der Bridge (aus dem Discovery-Suchlauf)
     * @throws IOException bei Fehlern beim Senden der Antwort
     */
    public void pairBridge(com.sun.net.httpserver.HttpExchange exchange, String bridgeId) throws IOException {
        hueBridgeController.pairBridge(exchange, bridgeId, () -> {
            // Callback nach erfolgreichem Pairing: Registriere Event-Stream-Manager
            try {
                HueModuleEventStreamManager moduleEventStreamManager = 
                    new HueModuleEventStreamManager(bridgeId, actionManager, databaseManager);
                eventStreamManager.registerModuleEventStreamManager(moduleEventStreamManager);
                logger.info("EventStreamManager für Bridge {} registriert", bridgeId);
            } catch (Exception e) {
                logger.error("Fehler beim Registrieren des EventStreamManagers für Bridge {}", bridgeId, e);
            }
        });
    }
    
    /**
     * Sucht alle Geräte einer gepaarten Hue Bridge.
     * Delegiert den Aufruf an HueDiscover.
     * 
     * @param exchange HTTP-Exchange für die Antwort
     * @param bridgeId Die deviceId der Bridge (aus dem Discovery-Suchlauf)
     * @throws IOException bei Fehlern beim Senden der Antwort
     */
    public void discoverDevices(com.sun.net.httpserver.HttpExchange exchange, String bridgeId) throws IOException {
        logger.info("Suche nach Geräten für Bridge: {}", bridgeId);
        
        try {
            List<Device> devices = hueDiscover.discoverDevices(bridgeId);
            
            if (devices != null && !devices.isEmpty()) {
                String response = gson.toJson(devices);
                ApiRouter.sendResponse(exchange, 200, response);
            } else {
                ApiRouter.sendResponse(exchange, 200, gson.toJson(new java.util.ArrayList<>()));
            }
            
        } catch (Exception e) {
            logger.error("Fehler bei der Geräte-Suche für Bridge {}", bridgeId, e);
            ApiRouter.sendResponse(exchange, 500, 
                gson.toJson(Map.of("error", "Fehler bei der Geräte-Suche: " + e.getMessage())));
        }
    }
    
    /**
     * Setzt die Sensitivity eines Hue Motion Sensors.
     * 
     * @param deviceId Die ID des Motion Sensors
     * @param sensitivity Die neue Sensitivity (0-100)
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setSensitivity(String deviceId, int sensitivity) {
        logger.info("Setze Sensitivity für Hue Motion Sensor: {} auf {}", deviceId, sensitivity);
        
        // Validiere Sensitivity-Wert (0-100)
        if (sensitivity < 0 || sensitivity > 100) {
            return false;
        }
        
        // Lade Device aus der Datenbank
        Optional<Device> device = actionManager.getDevice(deviceId);
        
        if (device.isEmpty() || !(device.get() instanceof DeviceMotion)) {
            return false;
        }

        try{
            ((DeviceMotion) device.get()).setSensibility(sensitivity, true);
            return actionManager.saveDevice(device.get());
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Sensitivity für Hue Motion Sensor {}", deviceId, e);
            return false;
        }
    }
    
    /**
     * Schaltet ein Hue Light ein.
     * 
     * @param deviceId Die ID des Light Devices
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setOn(String deviceId) {
        logger.info("Schalte Hue Light ein: {}", deviceId);
        
        Optional<Device> deviceOpt = actionManager.getDevice(deviceId);
        
        if (deviceOpt.isEmpty() || !(deviceOpt.get() instanceof DeviceLight)) {
            return false;
        }

        try {
            ((DeviceLight) deviceOpt.get()).setOn(true);
            return actionManager.saveDevice(deviceOpt.get());
        } catch (Exception e) {
            logger.error("Fehler beim Einschalten des Hue Light {}", deviceId, e);
            return false;
        }
    }
    
    /**
     * Schaltet ein Hue Light aus.
     * 
     * @param deviceId Die ID des Light Devices
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setOff(String deviceId) {
        logger.info("Schalte Hue Light aus: {}", deviceId);
        
        Optional<Device> device = actionManager.getDevice(deviceId);
        
        if (device.isEmpty() || !(device.get() instanceof DeviceLight)) {
            return false;
        }

        try {
            // Verwende die Basis-Methode, die für alle Hue Light Typen funktioniert
            ((DeviceLight) device.get()).setOff(true);
            return actionManager.saveDevice(device.get());
        } catch (Exception e) {
            logger.error("Fehler beim Ausschalten des Hue Light {}", deviceId, e);
            return false;
        }
    }
    
    /**
     * Setzt die Helligkeit eines Hue Light.
     * 
     * @param deviceId Die ID des Light Devices
     * @param brightness Die neue Helligkeit (0-100)
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setBrightness(String deviceId, int brightness) {
        logger.info("Setze Helligkeit für Hue Light: {} auf {}", deviceId, brightness);
        
        // Validiere Brightness-Wert (0-100)
        if (brightness < 0 || brightness > 100) {
            return false;
        }
        
        Optional<Device> device = actionManager.getDevice(deviceId);
        
        if (device.isEmpty() || !(device.get() instanceof DeviceLightDimmer)) {
            return false;
        }

        try {
            // Verwende die Basis-Methode, die für HueLightDimmer, HueLightDimmerTemperature
            // und HueLightDimmerTemperatureColor funktioniert
            ((DeviceLightDimmer) device.get()).setBrightness(brightness, true);
            return actionManager.saveDevice(device.get());
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Helligkeit für Hue Light {}", deviceId, e);
            return false;
        }
    }
    
    /**
     * Setzt die Farbtemperatur eines Hue Light.
     * 
     * @param deviceId Die ID des Light Devices
     * @param temperature Die neue Farbtemperatur (in mirek)
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setTemperature(String deviceId, int temperature) {
        logger.info("Setze Farbtemperatur für Hue Light: {} auf {}", deviceId, temperature);
        
        Optional<Device> device = actionManager.getDevice(deviceId);
        
        if (device.isEmpty() || !(device.get() instanceof DeviceLightDimmerTemperature)) {
            return false;
        }

        try {
            // Verwende die Basis-Methode, die sowohl für HueLightDimmerTemperature 
            // als auch für HueLightDimmerTemperatureColor funktioniert
            ((DeviceLightDimmerTemperature) device.get()).setTemperature(temperature, true);
            return actionManager.saveDevice(device.get());
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Farbtemperatur für Hue Light {}", deviceId, e);
            return false;
        }
    }
    
    /**
     * Setzt die Farbe eines Hue Light.
     * 
     * @param deviceId Die ID des Light Devices
     * @param x Die x-Komponente der Farbe (0.0-1.0)
     * @param y Die y-Komponente der Farbe (0.0-1.0)
     * @return true wenn erfolgreich, false sonst
     */
    public boolean setColor(String deviceId, double x, double y) {
        logger.info("Setze Farbe für Hue Light: {} auf ({}, {})", deviceId, x, y);
        
        // Validiere x und y Werte (0.0-1.0)
        if (x < 0.0 || x > 1.0 || y < 0.0 || y > 1.0) {
            return false;
        }
        
        Optional<Device> device = actionManager.getDevice(deviceId);
        
        if (device.isEmpty() || !(device.get() instanceof DeviceLightDimmerTemperatureColor)) {
            return false;
        }

        try {
            ((DeviceLightDimmerTemperatureColor) device.get()).setColor(x, y, true);
            return actionManager.saveDevice(device.get());
        } catch (Exception e) {
            logger.error("Fehler beim Setzen der Farbe für Hue Light {}", deviceId, e);
            return false;
        }
    }
    

}

File: src/main/node/com/smarthome/backend/server/api/modules/hue/hueModuleManager.ts
import { logger } from "../../../../logger.js";
import type { DatabaseManager } from "../../../db/database.js";
import { HueBridgeDiscover } from "./hueBridgeDiscover.js";
import { HueDeviceDiscover } from "./hueDeviceDiscover.js";
import { HueDeviceController } from "./hueDeviceController.js";
import { HueEventStreamManager } from "./hueEventStreamManager.js";
import { HueLight } from "./devices/hueLight.js";
import { HueLightDimmer } from "./devices/hueLightDimmer.js";
import { HueLightDimmerTemperatureColor } from "./devices/hueLightDimmerTemperatureColor.js";
import { HueLightDimmerTemperature } from "./devices/hueLightDimmerTemperature.js";
import { HueLightLevelSensor } from "./devices/hueLightLevelSensor.js";
import { HueTemperatureSensor } from "./devices/hueTemperatureSensor.js";
import { HueMotionSensor } from "./devices/hueMotionSensor.js";
import { HueLightLevelMotionTemperature } from "./devices/hueLightLevelMotionTemperature.js";
import { DeviceLight } from "../../../../model/devices/DeviceLight.js";
import { DeviceLightDimmer } from "../../../../model/devices/DeviceLightDimmer.js";
import { DeviceLightDimmerTemperature } from "../../../../model/devices/DeviceLightDimmerTemperature.js";
import { DeviceLightDimmerTemperatureColor } from "../../../../model/devices/DeviceLightDimmerTemperatureColor.js";
import { DeviceMotion } from "../../../../model/devices/DeviceMotion.js";
import { Device } from "../../../../model/devices/Device.js";
import { HueBridgeDiscovered } from "./hueBridgeDiscovered.js";
import { HueDeviceDiscovered } from "./hueDeviceDiscovered.js";
import { HueEvent } from "./hueEvent.js";
import { ModuleManagerBridged } from "../moduleManagerBridged.js";
import { HueBridgeController } from "./hueBridgeController.js";
import { EventManager } from "../../../events/EventManager.js";
import { HUECONFIG } from "./hueModule.js";
import { DeviceType } from "../../../../model/devices/helper/DeviceType.js";
import { HueSwitchDimmer } from "./devices/hueSwitchDimmer.js";
import { DeviceManager } from "../../entities/devices/deviceManager.js";

export class HueModuleManager extends ModuleManagerBridged<HueEventStreamManager, HueBridgeController, HueDeviceController, HueEvent, Device, HueDeviceDiscover, HueDeviceDiscovered, HueBridgeController, HueBridgeDiscover, HueBridgeDiscovered> {

  constructor(
    databaseManager: DatabaseManager,
    deviceManager: DeviceManager,
    eventManager: EventManager
  ) {
    const deviceController = new HueDeviceController(databaseManager);
    const bridgeController = new HueBridgeController(databaseManager);
    const bridgeDiscover = new HueBridgeDiscover(databaseManager);
    const deviceDiscover = new HueDeviceDiscover(databaseManager);
    
    super(
      databaseManager,
      deviceManager,
      eventManager,
      deviceController,
      deviceDiscover,
      bridgeController,
      bridgeDiscover
    );
  }

  public getModuleId(): string {
    return HUECONFIG.id;
  }
  protected getManagerId(): string {
    return HUECONFIG.managerId;
  }

  protected createEventStreamManager(): HueEventStreamManager {
    return new HueEventStreamManager(this.getManagerId(), this.bridgeController, this.deviceManager, this.databaseManager);
  }

  async getBridges() {
    return this.bridgeDiscover.getBridges()
      .filter(bridge => bridge.isPaired === true)
      .map(bridge => ({
      id: bridge.id,
      name: bridge.name,
      address: bridge.address,
      port: bridge.port,
      modelId: bridge.modelId,
      devices: bridge.devices,
      swVersion: bridge.swVersion,
      isPaired: bridge.isPaired,
    }));
  }

  async discoverBridges() {
    return this.bridgeDiscover.discover(10);
  }

  async pairBridge(bridgeId: string, _payload: Record<string, unknown>): Promise<boolean> {
    const bridge = this.bridgeDiscover.getBridge(bridgeId);
    if (!bridge) return false;
    const pairedBridge = await this.bridgeController.pairBridge(bridge);
    if (pairedBridge) {
       this.initialiseEventStreamManager();
       return true;
    }
    return false;
  }

  async discoverDevicesForBridge(bridgeId: string) {
    const devices = await this.deviceDiscover.discoverDevices(bridgeId);
    this.deviceManager.saveDevices(devices);
    return devices;
  }

  async setSensitivity(deviceId: string, sensitivity: number): Promise<boolean> {
    if (sensitivity < 0 || sensitivity > 100) return false;
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceMotion)) return false;
    await device.setSensibility(sensitivity, true);
    return this.deviceManager.saveDevice(device);
  }

  async setOn(deviceId: string): Promise<boolean> {
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceLight)) return false;
    await device.setOn(true);
    return this.deviceManager.saveDevice(device);
  }

  async setOff(deviceId: string): Promise<boolean> {
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceLight)) return false;
    await device.setOff(true);
    return this.deviceManager.saveDevice(device);
  }

  async setBrightness(deviceId: string, brightness: number): Promise<boolean> {
    if (brightness < 0 || brightness > 100) return false;
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceLightDimmer)) return false;
    await device.setBrightness(brightness, true);
    return this.deviceManager.saveDevice(device);
  }

  async setTemperature(deviceId: string, temperature: number): Promise<boolean> {
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceLightDimmerTemperature)) return false;
    await device.setTemperature(temperature, true);
    return this.deviceManager.saveDevice(device);
  }

  async setColor(deviceId: string, x: number, y: number): Promise<boolean> {
    if (x < 0 || x > 1 || y < 0 || y > 1) return false;
    const device = this.deviceManager.getDevice(deviceId);
    if (!device) return false;
    if (!(device instanceof DeviceLightDimmerTemperatureColor)) return false;
    await device.setColor(x, y, true);
    return this.deviceManager.saveDevice(device);
  }

  async convertDeviceFromDatabase(device: Device): Promise<Device | null> {
    if (device.moduleId !== this.getModuleId()) {
      return null;
    }

    const deviceType = device.type as DeviceType;
    let convertedDevice: Device | null = null;

    switch (deviceType) {
      case DeviceType.LIGHT:
        const hueLight = new HueLight();
        Object.assign(hueLight, device);
        await hueLight.updateValues();
        convertedDevice = hueLight;
        break;
      case DeviceType.LIGHT_DIMMER:
        const hueLightDimmer = new HueLightDimmer();
        Object.assign(hueLightDimmer, device);
        await hueLightDimmer.updateValues();
        convertedDevice = hueLightDimmer;
        break;
      case DeviceType.LIGHT_DIMMER_TEMPERATURE:
        const hueLightDimmerTemperature = new HueLightDimmerTemperature();
        Object.assign(hueLightDimmerTemperature, device);
        await hueLightDimmerTemperature.updateValues();
        convertedDevice = hueLightDimmerTemperature;
        break;
      case DeviceType.LIGHT_DIMMER_TEMPERATURE_COLOR:
        const hueLightDimmerTemperatureColor = new HueLightDimmerTemperatureColor();
        Object.assign(hueLightDimmerTemperatureColor, device);
        await hueLightDimmerTemperatureColor.updateValues();
        convertedDevice = hueLightDimmerTemperatureColor;
        break;
      case DeviceType.MOTION:
        const hueMotionSensor = new HueMotionSensor();
        Object.assign(hueMotionSensor, device);
        await hueMotionSensor.updateValues();
        convertedDevice = hueMotionSensor;
        break;
      case DeviceType.TEMPERATURE:
        const hueTemperatureSensor = new HueTemperatureSensor();
        Object.assign(hueTemperatureSensor, device);
        await hueTemperatureSensor.updateValues();
        convertedDevice = hueTemperatureSensor;
        break;
      case DeviceType.LIGHT_LEVEL:
        const hueLightLevelSensor = new HueLightLevelSensor();
        Object.assign(hueLightLevelSensor, device);
        await hueLightLevelSensor.updateValues();
        convertedDevice = hueLightLevelSensor;
        break;
      case DeviceType.MOTION_LIGHT_LEVEL_TEMPERATURE:
        const hueLightLevelMotionTemperature = new HueLightLevelMotionTemperature();
        Object.assign(hueLightLevelMotionTemperature, device);
        await hueLightLevelMotionTemperature.updateValues();
        convertedDevice = hueLightLevelMotionTemperature;
        break;
      case DeviceType.SWITCH_DIMMER:
        const hueSwitchDimmer = new HueSwitchDimmer();
        Object.assign(hueSwitchDimmer, device);
        await hueSwitchDimmer.updateValues();
        convertedDevice = hueSwitchDimmer;
        break;
    }

    return convertedDevice;
  }

  async initializeDeviceControllers(): Promise<void> {
    const devices = this.deviceManager.getDevicesForModule(this.getModuleId());
    
    for (const device of devices) {
        if (device instanceof HueLight || 
            device instanceof HueLightDimmer || 
            device instanceof HueLightDimmerTemperatureColor || 
            device instanceof HueLightDimmerTemperature || 
            device instanceof HueLightLevelSensor || 
            device instanceof HueTemperatureSensor || 
            device instanceof HueMotionSensor ||
            device instanceof HueSwitchDimmer) {
          device.setHueDeviceController(this.deviceController);
      }
    }
  }
}

